docs/design/federated-services.md

```diff
@@ -0,0 +1,490 @@
+# Kubernetes Cluster Federation (a.k.a. "Ubernetes")
+## Cross-cluster Load Balancing and Service Discovery 
+### Requirements and System Design
+### by Quinton Hoole, Dec 3 2015
+
+## Requirements
+
+### Discovery, Load-balancing and Failover
+
+1. **Internal discovery and connection**: Pods/containers (running in
+   a Kubernetes cluster) must be able to easily discover and connect
+   to endpoints for Kubernetes services on which they depend in a
+   consistent way, irrespective of whether those services exist in a
+   different kubernetes cluster within the same cluster federation.
+   Hence-forth referred to as "cluster-internal clients", or simply
+   "internal clients".
+1. **External discovery and connection**: External clients (running
+   outside a Kubernetes cluster) must be able to discover and connect
+   to endpoints for Kubernetes services on which they depend.
+   1. **External clients predominantly speak HTTP(S)**: External
+      clients are most often, but not always, web browsers, or at
+      least speak HTTP(S) (TBD list notable exceptions.  DNS servers?
+      SIP servers?  Database Servers?)
+1. **Find the "best" endpoint:** Upon initial discovery and
```

Comment:
Are we going to decide all of this on behalf of the user? or just expose it somehow. I don't think we can do the former because only the user is aware of things like write propogation latency/consistency semantics of their app. If we're going to expose it, how (other than just saying here are the good old loadbalancing algorithms gce/aws/nginx supports, pick one)? are you going to do something like profiles (cost effective, maximized availability etc)?

This feels complicated enough to stay away from till v1 is out and we have feedback.